File: tobas_gcs/src/tobas_gcs/apps/mission_planner/map_widget.py
import os.path as osp
from overrides import override
from typing import Tuple, List, Tuple, Dict, Type
from PyQt5.QtCore import (
    Qt,
    QObject,
    QVariant,
    QModelIndex,
    pyqtSignal,
    pyqtProperty,  # VSCodeで解析できないが存在する
    QStandardPaths,
    QUrl,
    QAbstractListModel,
)
from PyQt5.QtQuickWidgets import QQuickWidget
from PyQt5.QtPositioning import QGeoCoordinate


class AbstractListModel(QAbstractListModel):
    FIELDS: List[Tuple[str, Type]] = []

    def __init__(self) -> None:
        super().__init__()
        self._list: List[Tuple] = []

    @override
    def rowCount(self, _: QModelIndex = QModelIndex()) -> int:
        return len(self._list)

    @override
    def data(self, index: QModelIndex, role: Qt.ItemDataRole) -> QVariant:
        return self._list[index.row()][role - Qt.UserRole]

    @override
    def roleNames(self) -> Dict[Qt.ItemDataRole, bytes]:
        res = dict()
        for i, (name, _) in enumerate(self.FIELDS):
            res[Qt.UserRole + i] = name.encode()
        return res

    def add(self, *args) -> None:
        assert len(args) == len(self.FIELDS)
        for arg, (name, type_) in zip(args, self.FIELDS):
            assert isinstance(arg, type_), f"{arg}, {name}, {type_}"

        self.beginInsertRows(QModelIndex(), self.rowCount(), self.rowCount())
        self._list.append(args)
        self.endInsertRows()

    def clear(self) -> None:
        self.beginRemoveRows(QModelIndex(), 0, self.rowCount() - 1)
        self._list.clear()
        self.endRemoveRows()


class ImageModel(AbstractListModel):
    FIELDS = [("coordinate", QGeoCoordinate), ("source", QUrl)]


class WaypointModel(AbstractListModel):
    FIELDS = [("index", int), ("coordinate", QGeoCoordinate), ("acceptance_radius", float), ("marker_color", str)]


class LineModel(AbstractListModel):
    FIELDS = [("latitude_1", float), ("longitude_1", float), ("latitude_2", float), ("longitude_2", float)]


class SystemInfo(QObject):
    """QMLのコンストラクタ引数．"""

    _ = pyqtSignal()

    def __init__(self, parent=None):
        super(SystemInfo, self).__init__(parent)
        self._home_dir = QStandardPaths.writableLocation(QStandardPaths.HomeLocation)

    @pyqtProperty(str, notify=_)
    def homeDirectory(self):
        return self._home_dir


class MapWidget(QQuickWidget):
    """
    Open Street Mapを埋め込んだウィジェット．
    cf. https://stackoverflow.com/questions/36141170/creating-and-adding-mapquickitem-to-map-in-pyqt
    """

    def __init__(self) -> None:
        super().__init__(resizeMode=QQuickWidget.SizeRootObjectToView)  # リサイズモードの指定が必須

        # モデルオブジェクトの設定
        # QMLファイルのロード前に行う必要がある
        self._waypoint = WaypointModel()
        self._line = LineModel()
        self.rootContext().setContextProperty(WaypointModel.__name__, self._waypoint)
        self.rootContext().setContextProperty(LineModel.__name__, self._line)

        # コンストラクタ引数を設定
        system_info = SystemInfo()
        self.rootContext().setContextProperty(SystemInfo.__name__, system_info)

        # QMLファイルを読み込む
        qml_path = osp.join(osp.dirname(__file__), "qml/Map.qml")
        self.setSource(QUrl.fromLocalFile(qml_path))

        # シグナル
        self.waypoint_moved: pyqtSignal = self.rootObject().waypointMoved

    def clear(self) -> None:
        self._waypoint.clear()
        self._line.clear()

    def add_waypoint(self, index: int, coord: QGeoCoordinate, acceptance_radius: float, marker_color: str) -> None:
        self._waypoint.add(index, coord, acceptance_radius, marker_color)

    def add_line(self, latitude_1: float, longitude_1: float, latitude_2: float, longitude_2: float) -> None:
        self._line.add(latitude_1, longitude_1, latitude_2, longitude_2)

    def get_center(self) -> Tuple[float, float]:
        map = self.rootObject().findChild(QObject, "map")
        center: QGeoCoordinate = map.property("center")
        return center.latitude(), center.longitude()

    def set_center(self, latitude: float, longitude: float) -> None:
        assert -90 <= latitude <= 90
        assert -180 <= longitude <= 180

        self.rootObject().setCenter.emit(latitude, longitude)

File: tobas_gui/tobas_control_system/src/mission_planner/map_widget.cpp
#include <filesystem>
#include <ament_index_cpp/get_package_share_directory.hpp>
#include <QtQuick/QQuickItem>
#include <QtQml/QQmlContext>

#include <tobas_std_tools/check.hpp>

#include "tobas_control_system/constants.hpp"
#include "tobas_control_system/mission_planner/map_widget.hpp"
#include "tobas_control_system/mission_planner/system_info.hpp"

namespace fs = std::filesystem;

namespace gui
{
namespace gcs
{
MapWidget::MapWidget()
{
  // サイズポリシーとリサイズモードの指定 (しないとウィジェットが潰れてしまう)
  setSizePolicy(QSizePolicy::Preferred, QSizePolicy::Preferred);
  setResizeMode(QQuickWidget::SizeRootObjectToView);

  // モデルオブジェクトの設定
  // QMLファイルのロード前に行う必要がある
  waypoint_ = new map::WaypointModel();
  line_ = new map::LineModel();
  rootContext()->setContextProperty(waypoint_->modelName(), waypoint_);
  rootContext()->setContextProperty(line_->modelName(), line_);

  // コンストラクタ引数を設定
  const auto system_info = new SystemInfo();
  rootContext()->setContextProperty(system_info->modelName(), system_info);

  // QMLを読み込む
  const fs::path pkg_path(ament_index_cpp::get_package_share_directory(kPkgName));
  const auto qml_path = pkg_path / "qml/Map.qml";
  setSource(QUrl::fromLocalFile(QString::fromStdString(qml_path)));

  // Connection
  connect(rootObject(), SIGNAL(waypointMoved(int, double, double)), this, SLOT(onWaypointMoved(int, double, double)));
}

void MapWidget::clear()
{
  waypoint_->clear();
  line_->clear();
}

void MapWidget::addWaypoint(int index, const QGeoCoordinate& coord, double acceptance_radius, const QString& marker_color)
{
  waypoint_->add(index, coord, acceptance_radius, marker_color);
}

void MapWidget::addLine(double latitude_1, double longitude_1, double latitude_2, double longitude_2)
{
  line_->add(latitude_1, longitude_1, latitude_2, longitude_2);
}

QGeoCoordinate MapWidget::getCenter() const
{
  return getMapObject()->property("center").value<QGeoCoordinate>();
}

QGeoCoordinate MapWidget::getArrowPosition() const
{
  return getArrowObject()->property("coordinate").value<QGeoCoordinate>();
}

double MapWidget::getArrowRotation() const
{
  return getArrowRotationObject()->property("angle").value<double>();
}

void MapWidget::setMapCenter(double latitude, double longitude)
{
  QMetaObject::invokeMethod(rootObject(), "setMapCenter", Q_ARG(double, latitude), Q_ARG(double, longitude));
}

void MapWidget::setArrowPosition(double latitude, double longitude)
{
  QMetaObject::invokeMethod(rootObject(), "setArrowPosition", Q_ARG(double, latitude), Q_ARG(double, longitude));
}

void MapWidget::setArrowRotation(double angle_deg)
{
  QMetaObject::invokeMethod(rootObject(), "setArrowRotation", Q_ARG(double, angle_deg));
}

QObject* MapWidget::getMapObject() const
{
  const auto res = rootObject()->findChild<QObject*>("map");
  TOBAS_CHECK(res);
  return res;
}

QObject* MapWidget::getArrowObject() const
{
  const auto res = rootObject()->findChild<QObject*>("arrow");
  TOBAS_CHECK(res);
  return res;
}

QObject* MapWidget::getArrowRotationObject() const
{
  const auto res = rootObject()->findChild<QObject*>("arrowRotation");
  TOBAS_CHECK(res);
  return res;
}

void MapWidget::onWaypointMoved(int index, double latitude, double longitude)
{
  Q_EMIT waypointMoved(index, latitude, longitude);
}
}  // namespace gcs
}  // namespace gui
